core/src/main/java/org/apache/calcite/runtime/CompressionFunctions.java

```diff
@@ -0,0 +1,52 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to you under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.calcite.runtime;
+
+import org.apache.calcite.avatica.util.ByteString;
+
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.nio.ByteOrder;
+import java.nio.charset.Charset;
+import java.util.zip.DeflaterOutputStream;
+
+/**
+ * A collection of functions used in compression and decompression.
+ */
+public class CompressionFunctions {
+
+  private CompressionFunctions() {
+  }
+
+  public static ByteString compress(String data) {
+    try {
+      ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
+      ByteBuffer dataLength = ByteBuffer.allocate(4);
+      dataLength.order(ByteOrder.LITTLE_ENDIAN);
+      dataLength.putInt(data.length());
+      outputStream.write(dataLength.array());
+      DeflaterOutputStream inflaterStream = new DeflaterOutputStream(outputStream);
+      inflaterStream.write(data.getBytes(Charset.defaultCharset()));
+      inflaterStream.close();
```

Comment:
The MySQL says that
>  This function requires MySQL to have been compiled with a compression library such as zlib
Should we follow that ?

Reply:
I have validated the functionality by executing same function query on MySQL server.  Please let me know if I have misunderstood the question.

Reply:
From current code, it seems that there is no compression really, it's equivalent to `CAST(varchar as varbinary)` except that there is a header to mark the bytes length.

Reply:
@danny0405 Seems like there is some confusion, The current implementation does compression please refer test cases and attached MySQL output:
`mysql> select hex(compress(''));
+-------------------+
| hex(compress('')) |
+-------------------+
|                   |
+-------------------+
1 row in set (0.00 sec)

mysql> select hex(COMPRESS(REPEAT('a',1000)));
+--------------------------------------------+
| hex(COMPRESS(REPEAT('a',1000)))            |
+--------------------------------------------+
| E8030000789C4B4C1C05A360140C770000F9D87AF8 |
+--------------------------------------------+
1 row in set (0.00 sec)

mysql> select hex(compress(REPEAT('a',16)));
+--------------------------------+
| hex(compress(REPEAT('a',16)))  |
+--------------------------------+
| 10000000789C4B4C44050033980611 |
+--------------------------------+
1 row in set (0.00 sec)

mysql> select hex(COMPRESS('sample'));
+--------------------------------------+
| hex(COMPRESS('sample'))              |
+--------------------------------------+
| 06000000789C2B4ECC2DC849050008DE0283 |
+--------------------------------------+
1 row in set (0.00 sec)

mysql> select hex(COMPRESS('example'));
+----------------------------------------+
| hex(COMPRESS('example'))               |
+----------------------------------------+
| 07000000789C4BAD48CC2DC84905000BC002ED |
+----------------------------------------+
1 row in set (0.00 sec)`

Reply:
My fault, i didn't notice that there is a `DeflaterOutputStream`.

Reply:
No problem :)